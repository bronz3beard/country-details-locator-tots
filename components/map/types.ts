import { CSSProperties } from 'react';
import { LayerFeatureSource } from '~/lib/map-box/types';

export type InitialMapConfig = {
  lat: number;
  lng: number;
  source: LayerFeatureSource;
  zoom?: number;
  pitch?: number;
  bearing?: number;
  coordinates?: [number, number];
  styles?: CSSProperties;
};
