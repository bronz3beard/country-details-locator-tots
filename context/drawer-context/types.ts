export enum AppDrawers {
  MAP_SIDEBAR
}

export type DrawerValues = {
  [index: string]: boolean;
};
