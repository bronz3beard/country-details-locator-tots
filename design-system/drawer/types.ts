export enum DrawerAnchor {
  right,
  left
}
