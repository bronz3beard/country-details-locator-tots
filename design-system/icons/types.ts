export type SVGIconProps = {
  fill: string;
  width: number;
  height: number;
  color: string;
  'aria-hidden': string;
  viewBox: string;
  xmlns: string;
  className: string;
};

export type IconProps = {
  name: string;
  size?: number;
  color?: string;
  className?: string;
};
