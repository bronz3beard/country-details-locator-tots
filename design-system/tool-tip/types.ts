export enum ToolTipDirection {
  TOP,
  RIGHT,
  BOTTOM,
  LEFT
}
