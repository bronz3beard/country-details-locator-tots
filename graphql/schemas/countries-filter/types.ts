export type CountriesFilterQueryVariables = {
  query?: string | null;
};

export type CountriesFilterQuery = {
  countries: {
    nodes: Array<{
      code: string;
      name: string;
      currency: string;
    }>;
  };
};
