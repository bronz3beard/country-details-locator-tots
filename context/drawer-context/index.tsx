'use client';
import { Children, ReactElement, ReactNode, cloneElement, createContext, useMemo } from 'react';
import useDrawer, { UseDrawer } from '~/design-system/drawer/use-drawer';

export interface DrawerContextProps extends UseDrawer {}

export const ContextDrawer = createContext<DrawerContextProps>({
  drawerRef: null,
  drawerCloseButtonRef: null,
  showDrawer: {},
  handleOpenDrawer: () => {
    return;
  },
  handleCloseDrawer: () => {
    return;
  }
});

export const ProviderDrawer = ContextDrawer.Provider;

export const DrawerProvider = ({ children }: { children: ReactNode }) => {
  const { drawerRef, showDrawer, drawerCloseButtonRef, handleOpenDrawer, handleCloseDrawer } =
    useDrawer();

  const drawerApi: UseDrawer = useMemo(
    () => ({
      drawerRef,
      showDrawer,
      drawerCloseButtonRef,
      handleOpenDrawer,
      handleCloseDrawer
    }),
    [drawerRef, showDrawer, drawerCloseButtonRef, handleOpenDrawer, handleCloseDrawer]
  );

  return (
    <ProviderDrawer value={drawerApi}>
      {Children.count(children) > 1
        ? Children.map(children, (child) => cloneElement(child as unknown as ReactElement))
        : children}
    </ProviderDrawer>
  );
};

export default ContextDrawer;
