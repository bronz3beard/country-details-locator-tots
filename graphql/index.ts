import apolloClient from '~/lib/apolloClient';
import { assertIsTrue } from '~/utils/value-checker-utils';
import { handleQueryResult } from './helpers';
import { countryCodeFilterQuery } from './schemas/countries-filter/country-code-query';
import { countryNameFilterQuery } from './schemas/countries-filter/country-name-query';
import {
  CountriesFilterQuery,
  CountriesFilterQueryVariables
} from './schemas/countries-filter/types';
import { countryDetailsQuery } from './schemas/country-details.query';

export async function searchCountries({ query }: CountriesFilterQueryVariables) {
  const [codeQueryResult, nameQueryResult] = await Promise.all([
    apolloClient.query<CountriesFilterQuery>({
      query: countryCodeFilterQuery,
      variables: { query }
    }),
    apolloClient.query<CountriesFilterQuery>({
      query: countryNameFilterQuery,
      variables: { query: `.*${query}.*` } // Example regex to match names containing the query
    })
  ]);
  const codeResult = handleQueryResult(codeQueryResult);
  const nameResult = handleQueryResult(nameQueryResult);
  const codeNodes = codeResult.data?.countries?.nodes ?? [];
  const nameNodes = nameResult.data?.countries?.nodes ?? [];

  // Combine results and use a Set to ensure uniqueness based on `code`
  const combinedResults = [...codeNodes, ...nameNodes];

  const uniqueResults = Array.from(new Map(combinedResults.map((c) => [c.code, c])).values());

  return uniqueResults;
}

export async function getCountryDetailsByCountryCode({ query }: CountriesFilterQueryVariables) {
  const result = await apolloClient.query<CountriesFilterQuery>({
    query: countryDetailsQuery,
    variables: {
      query // or "United States"
    }
  });
  assertIsTrue(!!result);

  console.log('ðŸš€ ~ getCountryDetailsByCountryCode ~ result:', result);
  return handleQueryResult(result);
}
