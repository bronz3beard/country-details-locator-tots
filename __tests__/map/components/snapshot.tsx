import { act, render } from '@testing-library/react';
import { useRouter } from 'next/navigation';
import { assertSnapshot } from '~/__tests__/utils/assert-snapshot';
import { mockFunction } from '~/__tests__/utils/mock-function';
import renderPropComponent from '~/__tests__/utils/mock-render-props-component';
import { mockUseEffect } from '~/__tests__/utils/mock-use-effect';
import { mockUseState } from '~/__tests__/utils/mock-use-state';
import DetailsCard from '~/app/map/components/details-card';
import MapPageContainer from '~/app/map/components/map-page-container';
import ConditionalWrapper from '~/components/conditional-wrapper';
import PageLoader from '~/components/loaders/page-loader';
import { DrawerProvider } from '~/context/drawer-context';
import { CountryDetails, CountryDetailsQueryReturnData } from '~/graphql/types';

const mockDataDetail: CountryDetails = {
  country: {
    awsRegion: 'us-east-2',
    capital: 'Washington D.C.',
    code: 'US',
    continent: {
      name: 'North America'
    },
    currencies: ['USD', 'USN', 'USS'],
    currency: 'USD,USN,USS',
    emoji: 'ðŸ‡ºðŸ‡¸',
    emojiU: 'U+1F1FA U+1F1F8',
    languages: [
      {
        name: 'English',
        native: 'English'
      }
    ],
    name: 'United States of America',
    native: 'United States',
    phone: '1',
    phones: ['1'],
    states: [
      {
        name: 'Alabama',
        code: 'AL'
      }
    ],
    subdivisions: []
  }
};
const mockData: CountryDetailsQueryReturnData = [mockDataDetail];

mockFunction(useRouter);

jest.mock('mapbox-gl', () => ({
  Map: jest.fn(() => ({ on: jest.fn(), remove: jest.fn() })),
  LngLatBounds: jest.fn()
}));

jest.mock('~/components/map/use-init-map', () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue({
    mapBox: null,
    mapContainer: {}
  })
}));

jest.mock('~/hooks/use-map-features', () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue({
    featureList: [
      {
        Country: 'United States of America',
        'ISO Code': 'US',
        Latitude: 37.0902,
        Longitude: -95.5022
      }
    ],
    filterFeatures: jest.fn()
  })
}));

jest.mock('~/hooks/use-dialog', () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue({
    openDialog: true,
    handleOpenDialog: jest.fn(),
    handleDialogClose: jest.fn()
  })
}));

jest.mock('~/hooks/use-search-query-params', () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue({
    searchFilter: '',
    handleClearSearchQueryParam: jest.fn()
  })
}));

jest.mock('~/components/loaders/page-loader', () => 'PageLoader');
jest.mock('~/design-system/navbar', () => 'Navbar');
jest.mock('~/components/map/map-box', () => 'MapBox');
jest.mock('~/components/sidebar-menu', () => 'SidebarMenu');
jest.mock('~/design-system/drawer', () => 'Drawer');

describe('MapPageContainer', () => {
  beforeEach(() => {
    jest.resetModules();
    jest.clearAllMocks();
  });

  it('renders correctly and matches snapshot', async () => {
    const mockState = {
      ...mockDataDetail
    };

    await act(async () => {
      mockUseState({ isExpanded: true });

      mockUseState(mockState);
      mockUseEffect();
    });

    const element = renderPropComponent({
      Component: (
        <ConditionalWrapper conditional={mockState}>
          {({ value }) => {
            const country = value.country;
            return (
              <DetailsCard
                {...{
                  name: country.name,
                  states: country.states,
                  flag: `${country.emoji}`,
                  capital: country.capital,
                  currency: country.currency,
                  languages: country.languages,
                  continent: country.continent.name,
                  subdivisions: country.subdivisions
                }}
              />
            );
          }}
        </ConditionalWrapper>
      ),
      props: { conditional: mockState },
      testId: 'test-details-card'
    });

    expect(element).toBeInTheDocument();

    const { container } = render(
      <DrawerProvider>
        <MapPageContainer countries={mockData} loading={false} />
      </DrawerProvider>
    );

    assertSnapshot(container);
  });

  it('renders loading state and matches snapshot', async () => {
    const jsx = MapPageContainer({ countries: null, loading: true });

    await act(async () => {
      render(jsx);

      const { container } = render(<PageLoader />);
      expect(container).toBeInTheDocument();
    });

    assertSnapshot(jsx);
  });
});
