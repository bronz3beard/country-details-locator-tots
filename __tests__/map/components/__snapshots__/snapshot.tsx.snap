// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapPageContainer renders correctly and matches snapshot 1`] = `
<div>
  <navbar />
  <drawer
    anchor="1"
    drawer="0"
    leftcloselimit="md:-translate-x-64 -translate-x-80"
    width="w-80"
  />
  <div
    class="flex h-[calc(100vh-54px)] w-screen overflow-hidden bg-gray-100/50 bg-white bg-clip-border text-gray-700"
  >
    <div
      class="h-full w-full"
    >
      <mapbox
        mapcontainer="[object Object]"
      />
    </div>
  </div>
</div>
`;

exports[`MapPageContainer renders loading state and matches snapshot 1`] = `<PageLoader />`;
